server/src/main/java/org/apache/iotdb/db/metadata/mtree/traverser/Traverser.java

```diff
@@ -0,0 +1,210 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.metadata.mtree.traverser;
+
+import org.apache.iotdb.db.exception.metadata.IllegalPathException;
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.db.metadata.PartialPath;
+import org.apache.iotdb.db.metadata.mnode.IMNode;
+import org.apache.iotdb.db.metadata.mnode.MeasurementMNode;
+import org.apache.iotdb.db.metadata.template.Template;
+import org.apache.iotdb.tsfile.write.schema.IMeasurementSchema;
+
+import java.util.regex.Pattern;
+
+import static org.apache.iotdb.db.conf.IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD;
+import static org.apache.iotdb.db.conf.IoTDBConstant.ONE_LEVEL_PATH_WILDCARD;
+
+// This class defines the main traversal framework and declares some methods for result process
+// extension.
+public abstract class Traverser {
+
+  protected IMNode startNode;
+  protected String[] nodes;
+
+  // if isMeasurementTraverser, measurement in template should be processed
+  protected boolean isMeasurementTraverser = false;
+
+  // default false means fullPath pattern match
+  protected boolean isPrefixMatch = false;
+
+  public Traverser(IMNode startNode, PartialPath path) throws MetadataException {
+    String[] nodes = path.getNodes();
+    if (nodes.length == 0 || !nodes[0].equals(startNode.getName())) {
+      throw new IllegalPathException(
+          path.getFullPath(), path.getFullPath() + " doesn't start with " + startNode.getName());
+    }
+    this.startNode = startNode;
+    this.nodes = nodes;
+  }
+
+  public void traverse() throws MetadataException {
+    traverse(startNode, 0, 0);
+  }
+
+  /**
+   * The recursive method for MTree traversal.
+   *
+   * @param node current node that match the targetName in given path
+   * @param idx the index of targetName in given path
+   * @param level the level of current node in MTree
+   * @throws MetadataException some result process may throw MetadataException
+   */
+  protected void traverse(IMNode node, int idx, int level) throws MetadataException {
+
+    if (processMatchedMNode(node, idx, level)) {
+      return;
+    }
+
+    if (idx >= nodes.length - 1) {
+      if (nodes[nodes.length - 1].equals(MULTI_LEVEL_PATH_WILDCARD) || isPrefixMatch) {
+        processMultiLevelWildcard(node, idx, level);
+      }
+      return;
+    }
+
+    if (node.isMeasurement()) {
+      return;
+    }
+
+    String targetName = nodes[idx + 1];
+    if (MULTI_LEVEL_PATH_WILDCARD.equals(targetName)) {
+      processMultiLevelWildcard(node, idx, level);
+    } else if (targetName.contains(ONE_LEVEL_PATH_WILDCARD)) {
+      processOneLevelWildcard(node, idx, level);
+    } else {
+      processNameMatch(node, idx, level);
+    }
+  }
+
+  // process curNode that matches the targetName during traversal
+  private boolean processMatchedMNode(IMNode node, int idx, int level) throws MetadataException {
```

Comment:
add javadoc for return value or remove this method

Reply:
javadoc added.